release-20.2: kvserver: fix deadlock in Store.removeInitializedReplicaRaftMuLocked

erikgrinaker opened the pull request:
Backport 1/1 commits from #65781.
Backport 1/1 commits from #67006.

We should let this bake on master for a little bit, in case removing the channel buffer causes any trouble.

/cc @cockroachdb/release @cockroachdb/kv

---

When `Store.removeInitializedReplicaRaftMuLocked()` is finishing up
replica removal, it calls `replicaScanner.RemoveReplica()` to remove it
from the queue, which happens via a buffered channel. However, this call
(and thus the channel send) was done while still holding `Store.mu`,
which could deadlock if the channel buffer filled up and the
`replicaScanner` also needed to grab `Store.mu`.

Since there is no need for `Store.mu` to be held when calling
`replicaScanner.RemoveReplica()`, this patch unlocks it first. It also
removes the `replicaScanner.removed` channel buffer, to avoid hiding
these sorts of deadlocks in the first place.

Release note (bug fix): Fixed a deadlock that could occur when many
replicas were rapidly queued for removal.

erikgrinaker committed:
kvserver: fix deadlock in Store.removeInitializedReplicaRaftMuLocked

When `Store.removeInitializedReplicaRaftMuLocked()` is finishing up
replica removal, it calls `replicaScanner.RemoveReplica()` to remove it
from the queue, which happens via a buffered channel. However, this call
(and thus the channel send) was done while still holding `Store.mu`,
which could deadlock if the channel buffer filled up and the
`replicaScanner` also needed to grab `Store.mu`.

Since there is no need for `Store.mu` to be held when calling
`replicaScanner.RemoveReplica()`, this patch unlocks it first. It also
removes the `replicaScanner.removed` channel buffer, to avoid hiding
these sorts of deadlocks in the first place.

Release note (bug fix): Fixed a deadlock that could occur when many
replicas were rapidly queued for removal.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/65861)
<!-- Reviewable:end -->

erikgrinaker committed:
kvserver: fix data race on replicaScanner.stopper

In #65781, a race was introduced on `replicaScanner.stopper`, since it
is set by `Start()` and accessed by `RemoveReplica()` but `Start()` is
called asynchronously.

To avoid introducing a mutex around the stopper or refactoring the store
construction, this adds a hack that writes the stopper directly to
`replicaScanner.stopper` synchronously during `Store.Start()`.

Release note: None

erikgrinaker commented:
@tbg TFTR! Would you mind approving the corresponding 21.1 backport in #65859?
